fluid/neural_machine_translation/transformer/reader.py

```diff
@@ -263,95 +262,41 @@ def load_dict(dict_path, reverse=False):
                     word_dict[line.strip()] = idx
         return word_dict
 
-    def _sample_generator(self):
+    def batch_generator(self):
+        # global sort or global shuffle
+        beg = time.time()
         if self._sort_type == SortType.GLOBAL:
-            if not self._sorted:
-                self._sample_idxs.sort(
-                    key=lambda idx: max(len(self._src_seq_ids[idx]),
-                    len(self._trg_seq_ids[idx] if not self._only_src else 0))
-                )
-                self._sorted = True
+            infos = sorted(
+                self._sample_infos,
+                key=lambda x: max(x[1], x[2]) if not self._only_src else x[1])
         elif self._shuffle:
-            random.shuffle(self._sample_idxs)
-
-        for sample_idx in self._sample_idxs:
-            if self._only_src:
-                yield (self._src_seq_ids[sample_idx], )
-            else:
-                yield (self._src_seq_ids[sample_idx],
-                       self._trg_seq_ids[sample_idx][:-1],
-                       self._trg_seq_ids[sample_idx][1:])
-
-    def batch_generator(self):
-        pool = Pool(self._sample_generator, self._pool_size, True
-                    if self._sort_type == SortType.POOL else False)
-
-        def next_batch():
-            batch_data = []
-            max_len = -1
-            batch_max_seq_len = -1
-
-            while True:
-                sample = pool.next(look=True)
+            infos = self._sample_infos
+            self._random.shuffle(infos)
+        else:
+            infos = self._sample_infos
```

Comment:
We may also need local sort to sort by groups as SortType.POOL indicates. It is better to support this.